src/routing.rs

```diff
@@ -214,7 +200,12 @@ pub fn setup_popstate_listener<Ms>(
             if let Some(routing_msg) = routes(url) {
                 update(routing_msg);
             }
-        };
+        } else {
+            // `ev.state()` will be a null JsValue for the landing page.
+            if let Some(routing_msg) = routes(initial_url()) {
+                update(routing_msg);
+            }
+        }
```

Comment:
We should use `if`s as little as possible, especially `if let` is dangerous (as we found out because of the original issue).
And I don't like to respect DRY at all cost, but in this case we can reuse block with `update(routing_msg)`.

So I suggest to change
```rust
      if let Some(state_str) = ev.state().as_string() {
            let url: Url =
                serde_json::from_str(&state_str).expect("Problem deserializing popstate state");
            // Only update when requested for an update by the user.
            if let Some(routing_msg) = routes(url) {
                update(routing_msg);
            }
        } else {
            // `ev.state()` will be a null JsValue for the landing page.
            if let Some(routing_msg) = routes(initial_url()) {
                update(routing_msg);
            }
        }
```
To something like:
```rust
        // `ev.state()` will be a null `JsValue` for the landing page.
        let url = match ev.state().as_string() {
            Some(state_str) => serde_json::from_str(&state_str).expect("Problem deserializing popstate state"),
            None => initial_url()
        };
        // Only update when requested for an update by the user.
        if let Some(routing_msg) = routes(url) {
            update(routing_msg);
        }
```

Reply:
Interesting note about `if let` being dangerous, I've been pretty liberal with it but this issue is a good example of why exhaustive matching is good.

Reply:
Yeah, my colleagues thought that I'm weird when I was writing exhaustive switches also in Typescript 😃